### PR TITLE
Open text files with utf-8 encoding

### PR DESCRIPTION
**Pull request by dschuff**
Reads from text files return strings, which are unicode on python3.
On Windows the default encoding for opening text files is not utf-8, which causes decode failure in the presence of certain codepoints. A similar problem happens when encoding for writes.
This PR fixes all of the cases where we open text files with user-controlled content in `emcc.py`, `shared.py`, and `emscripten.py`. There are probably more cases we need to fix.

### CODE REVIEW

**Comment by dschuff** [comment]
that's an odd CI failure. any idea what the deal is?

**Comment by sbc100** [comment]
Can we no set the default somehow instead?

**Comment by dschuff** [comment]
oh, the problem is that you can't use the `encoding` kwarg in python2, and python2 sneaks into the build via CMake calling back into emcc without passing sys.executable.... so yeah this PR won't work.

**Comment by dschuff** [comment]
I haven't found a good way to set the default yet.

**Comment by dschuff** [comment]
(at least I haven't found a good way poking through the docs. that sounds like a stackoverflow kind of question though).

File: emcc.py [hunk 1]
```diff
@@ -655,13 +655,13 @@ def is_minus_s_for_emcc(args, i):
     if debug_configure:
       tempout = '/tmp/emscripten_temp/out'
       if not os.path.exists(tempout):
-        open(tempout, 'w').write('//\n')
+        open(tempout, 'w', encoding='utf-8').write('//\n')
```
Comment (kripken) [inline emcc.py]: this is a write, do we need to set the encoding here?

Comment (kripken) [reply emcc.py]: (a write of a trivial string)

**Comment by kripken** [comment]
For setting the default, maybe
```python
import locale
locale.setlocale(locale.LC_ALL, 'C.UTF-8')
```
..?

File: emcc.py (hunk 1 above)
Comment (dschuff) [reply emcc.py]: oh; no it only actually needs to be set if there could be any unicode written.

**Comment by dschuff** [comment]
that `setlocale` doesnt work: I get an error that the locale isn't supported (same when I try 'POSIX').

I have a workaround for running our tests, which is that the encoding can be overridden at python load time (https://www.python.org/dev/peps/pep-0540/). I think we maybe want to think a bit more before trying to override the locale for every emscripten invocation everywhere; e.g. if a user has source files encoded in cp1252 on their system already then that change will break them. We should look more into how JS and C++ files are expected to be encoded first.

**Comment by sbc100** [comment]
I don't think we reason cpp or h files using python though right?   clang is in charge of reading those.. 
  Only JS files are read by python under normal operation right?  (except in tests of course but can control that).

**Comment by dschuff** [comment]
We don't reason about them, but we do open them, at least in the test runner (e.g. `do_run_from_file` implements running tests from the filesystem by just opening them and passing them to `do_run`). Also plenty of places in emscripten process user-written JS (or come to think of it, fragments of user-written C++ e.g. EM_JS). The error happens anytime you need to decode the content into a string.

**Comment by kripken** [comment]
I guess we can document the `PYTHONUTF8=1` trick for our users. That might be ok enough for now as this is very rare in practice.

**Comment by dschuff** [comment]
If we want to do this longterm it probably also makes sense to put it in our shell and batch wrappers too. But we still probably want to be more careful in the code as well AFAICS JS should always be encoded as UTF-8 but the default behavior in MSVS is to save C++ source files with the default locale and not necessarily UTF-8. So users might possibly have legacy sources with non-UTF encodings, and if they want to add EM_JS to them they would break. So we might want to offer a way to change it; e.g. by supporting `-finput-charset` and/or `-fexec-charset`.

**Pull request closed without merging**